rfcs/ion_1_1/feature-inline_symbols.md

```diff
@@ -0,0 +1,755 @@
+	
+Consider a template representing a class called `com.example.project.Quux`:
+
+```js
+$ion_1_1
+
+// Define a template composed of several symbols:
+
+$ion_symbol_table::{
+  templates: [
+    com.example.project.Quux::{ // Template 1, a struct with 3 blanks
+	  foo: {#0},
+	  bar: {#0},
+	  baz: {#0}
+	}
+  ]
+}
+
+// Invoke template 1 with various different parameters
+
+{#1 1 2 3}
+{#1 4 5 6}
+{#1 7 8 9}
+
+// The above invocations are equivalent to:
+
+com.example.project.Quux::{
+  foo: 1,
+  bar: 2,
+  baz: 3,
+}
+com.example.project.Quux::{
+  foo: 4,
+  bar: 5,
+  baz: 6,
+}
+com.example.project.Quux::{
+  foo: 7,
+  bar: 8,
+  baz: 9,
+}
+```
+
+If we use inline symbol definitions for `foo`, `bar`, `baz`, and `com.example.project.Quux` in our
+template definition, we do not need to add them to the symbol table. By allocating a single template
+ID, we can produce a compact representation of a struct composed of several symbols without growing
+the symbol table.
+
+If the template is frequently used, it can be placed in a Shared Symbol Table.
+
+## Alternatives considered
+
+### Add some new encodings, but not others
+
+For example, add inline symbol structs but not inline symbol values. This would conserve type
+descriptor values and reduce the scope of the changes needed for Ion 1.1. However, it would
+eliminate a key feature for writers.
+
+Inline symbol definitions make it possible for writers to serialize a number of values without
+having to modify the active symbol table. This makes it possible to:
+1. Enforce restrictions on the size of the active symbol table (either by number of entries or data
+size) without having to resort to resetting it.
+2. Batch changes to the symbol table to avoid frequent LST appends.
+3. Guarantee that all changes to the symbol table happen at particular offsets. (For example,
+symbol table appends only appear every 8MB in a stream.) This makes it possible to skip-scan
+over huge streams while only checking at each 8MB offset to see if there are symbol table changes
+that need to be processed.
+
+### Encode maps, dicts, etc as something other than a struct
+
+For example, write maps out as lists of key/value pairs:
+
+```js
+[(foo, 1), (bar, 2), (baz, 3)]
+```
+
+or lists in which even-indexed entries are keys and odd-indexed entries are their associated
+values:
+
+```js
+[foo, 1, bar, 2, baz, 3]
+```
+
+Such solutions are implemented as a layer on top of the Ion data model itself, which presents
+additional challenges. All applications relying on a custom representation of a map must perform
+their own validation to guarantee that every key has an associated value and that each key is a
+string or symbol value. Tools such as [PartiQL](https://partiql.org/) that are designed to work with
+Ion will not recognize custom representations as maps, meaning that convenient syntax meant to work
+with key/value data will not function as expected.
```

Comment:
This is a weak argument, since there are a number of higher-level data types, notably set/bag, that are not well-supported by Ion and that require special handling at higher layers.

But really this section doesn't solve the problems noted, it just moves it around.